R/secchi_workflow.R

```diff
@@ -0,0 +1,189 @@
+    params$startDateLo = '1900-01-01'
+  }
+  
+  if(!('startDateHi' %in% names(params))){
+    params$startDateHi = format(Sys.Date(), '%Y-%m-%d')
+  }
+  
+  if('statecode' %in% names(params)){
+    params$statecode = paste0("US:",params$statecode)
+  }
+  
+  starts = seq(as.POSIXct(params$startDateLo), as.POSIXct(params$startDateHi), by=stride)
+  
+  if(length(starts) < 2){
+    ends   = as.POSIXct(params$startDateHi)
+  } else {
+    ends   = c(starts[2:(length(starts))], as.POSIXct(params$startDateHi))
+  }
+  
+  starts = format(starts, '%Y-%m-%d')
+  ends   = format(ends, '%Y-%m-%d')
+  
+  out = list()
+  
+  for(i in seq_along(starts)){
+    cat('Downloading ', starts[i], ' to ', ends[i], '\n')
+    params$startDateLo = starts[i]
+    params$startDateHi = ends[i]
+    
+    if(starts[[i]] != ends[[i]]){
+      chunk.call = do.call(retryWQP, params)
+    
+      if(!is.null(chunk.call)){
+        meta <- attr(chunk.call, "siteInfo") %>%
+          select(MonitoringLocationIdentifier, dec_lat_va, dec_lon_va, StateCode)
+        
+        chunk.call <- rename(chunk.call, 
+                             Date=ActivityStartDate, 
+                             value=ResultMeasureValue, 
+                             units=ResultMeasure.MeasureUnitCode) %>% 
+          left_join(meta, by="MonitoringLocationIdentifier") %>%
+          rename(wqx.id=MonitoringLocationIdentifier) %>%
+          select(Date, value, units, wqx.id, dec_lat_va, dec_lon_va, StateCode)
+        
+        out[[i]] <- chunk.call       
+      }
+    }
+  }
+  
+  if(length(out) > 0){
+    out.data <- bind_rows(out)
+    
+  } else {
+    out.data <- data.frame(Date = numeric(),
+                           value=numeric(),
+                           units = character(),
+                           wqx.id = character(),
+                           dec_lat_va = numeric(),
+                           dec_lon_va = numeric(),
+                           StateCode = character())
+  }
+  
+  return(out.data)
+
+}
+
+retryWQP <- function(..., retries=3){
+  
+  safeWQP = function(...){
+    result = tryCatch({
+      readWQPdata(...)
+    }, error = function(e) {
+      if(e$message == 'Operation was aborted by an application callback'){
+        stop(e)
+      }
+      return(NULL)
+    })
+    return(result)
+  }
+  retry = 1
+  while (retry < retries){
+    result = safeWQP(...)
+    if (!is.null(result)){
+      retry = retries
+    } else {
+      message('query failed, retrying')
+      retry = retry+1
+    }
+  }
+  return(result)
+}
+
+library(dplyr)
+library(dataRetrieval)
+library(lubridate)
+library(ggplot2)
+
+characteristicNames = c("Depth, Secchi disk depth", "Depth, Secchi disk depth (choice list)", "Secchi Reading Condition (choice list)", "Secchi depth", "Water transparency, Secchi disc")
+siteTypes = "Lake, Reservoir, Impoundment"
+
+lower.48 <- stateCd$STATE[1:50]
+lower.48 <- lower.48[!(stateCd$STUSAB[1:50] %in% c("AL","HI"))]
```

Comment:
these two lines should be commented out since they are only needed to reproduce the whole fig (and the rest of that code is commented out)